### ISSUE TITLE
Implement __getstate__/__setstate__ for dict classes

### ISSUE THREAD

**Issue by TMiguelT** [python-attrs/attrs#741]
Following on from #735. To get `attr.s` classes loading correctly from YAML, I see the path of least resistance is implementing a `__getstate__`/`__setstate__` for dict classes. 

Other options I have rejected are:
* Getting `ruamel.yaml` to use `__init__` instead of `__new__`: this would undoubtedly break other code that relies on the constructor not being called
* Implement my own `__getstate__`/`__setstate__` for each of my classes, or as a mixin: this would work, but why not push this implementation back into `attrs` so that others can benefit from it?
* Implement `to_yaml`/`from_yaml`, which are the `ruamel.yaml`-specific versions of `__getstate__`/`__setstate__`: this would also work, but `attr.s` classes don't need to specify how the YAML is actually written, just how to represent the classes as a dict. By using the generic `__getstate__`/`__setstate__`, we know that other serializers will be able to use these methods too

Currently, it's recommended that you don't use `getstate_setstate` without `slots`, but there's no particular reason why this must be the case. The current `__setstate__` implementation doesn't work for dict classes because it tries to zip the state with the attributes. If the state is a dictionary, that means zipping the keys of that dictionary with the class attributes, and using the keys as values. This obviously makes no sense for a dict class: https://github.com/python-attrs/attrs/blob/f132c07e556dde35d5f7ef30b736800eb08e9f3b/src/attr/_make.py#L848-L854

Because of this, we need a custom implementation for dict classes. However this implementation could be very simple, such as:

```python
def __setstate__(self, state: dict): 
    self.__dict__.update(state)
    self.__init__()
```

**Comment by Tinche** [comment]
Ok, I looked at the original issue and I can comment on this. I am not familiar with ruamel.yaml and familiar with yaml only on a basic level (I have written an awful lot of it for Kubernetes but don't know the ins and outs of the spec). Is ruamel.yaml your library?

The proper way to instantiate attrs classes (and dataclasses, for that matter) is to call the initializer, like you would do yourself in your code. This dance with `__new__` and then setting attributes reminds me a lot of Java frameworks from back in the day, and I personally think it's hacky and completely unnecessary in Python. So, I would advocate for ruamel.yaml to adopt the approach of instantiating classes the right way.

I also don't think the path of least resistance is carrying this `__setstate__`/`__getstate__` logic in attrs. If you harness the yaml library only for Python primitives, you can use another library like cattrs to structure this into classes. Adopting your example:

```
import attr
from typing import List, Optional
from ruamel.yaml import YAML
from cattr import structure

yaml=YAML(typ='safe')


@attr.s(auto_attribs=True)
class Command:
    command: List[str]
    docker_image: Optional[str] = None

print(structure(yaml.load(open("t.yaml").read()), Command))
```
```
>>> Command(command=['samtools'], docker_image=None)
```
Note that I had to adjust your yaml input by removing the first line.
```
command:
  - samtools
```
Doing it this way makes defaults work, works with any type of attrs classes (slots and dict), and allows you to switch to another serialization format trivially.

**Comment by TMiguelT** [comment]
>  Is ruamel.yaml your library?

No, it's a very popular library for YAML (de)serialisation in Python

> The proper way to instantiate attrs classes (and dataclasses, for that matter) is to call the initializer

Right, but a deserializer can't rely on it always being correct to call the initialiser. We need to tell it to do so for attrs classes, which is exactly what I do in my sample above.

> If you harness the yaml library only for Python primitives, you can use another library like cattrs to structure this into classes

I don't dislike this solution, but I've already committed to using tagged YAML, which specifically tags each YAML object literal with the corresponding Python class. I don't see why this shouldn't be supported as well.

**Comment by TMiguelT** [comment]
Also, to phrase this another way: `__setstate__` is broken for dict classes, although the API lets you do it. Why not let me fix it, and get better YAML compatibility along the way?

**Comment by TMiguelT** [comment]
Oh, actually I think `__setstate__` is totally broken for dictionary state? Even if I set my internal classes to `slots=True`, it still tries to zip the dict keys with the attribute names, resulting in a failed final state. Try it yourself:

```python
import attr
@attr.s(slots=True, auto_attribs=True)
class MyClass:
    some_field: str = ""
state = {'some_field': 'foo'}
instance = MyClass()
```
```python
>>> instance
MyClass(some_field='')
>>> instance.__setstate__(state)
>>> instance
MyClass(some_field='some_field')
```
`instance.some_field` should of course have the value `"foo"` here.

**Comment by Tinche** [comment]
According to the Python docs, `__setstate__` and `__getstate__` are interfaces for classes to support pickle? I guess ruamel is trying to piggyback on these. Pickling attrs classes works though, as far as I know.

**Comment by TMiguelT** [comment]
You're correct, it's piggybacking off a pickle interface.

> Pickling attrs classes works though, as far as I know.
 From my comments above, I don't believe this is true for non-slots classes (ie dict classes)

**Comment by Drino** [comment]
Hello again :)

TL;DR

* `__getstate__`/`__setstate__` work fine with `__dict__`-classes
* `__getstate__`/`__setstate__` aren't expected to be anymore complex than `def __getstate__(self): return self.__dict__` and `def __setstate__(self, state): self.__dict__.update(state)`. Basically, pickle dumps are not expected to be backward-compatible when your code changes (but for same code they expected to be python interpreter version compatible).
* Best approach for backward-compatibility is to mind backward-compatibility of your own classes where you need it.

It seems that you misunderstood the attrs-generated `__getstate__` and `__setstate__` in `__dict__`-classes, e.g.:
```python
import attr

@attr.dataclass(getstate_setstate=True)
class TestMe:
    s: str
    i: int = attr.ib(kw_only=True)

t = TestMe('Hello!', i=42)
t_new = TestMe.__new__(TestMe)
t_new.__setstate__(t.__getstate__())

print(t_new)  # TestMe(s='Hello!', i=42)
```

Sounds like your main problem (mentioned in the previous issue) was not the actual `__getstate__`/`__setstate__` implementation, but their backward-compatibility.

There is a couple of problems here, which makes it tricky to provide a good design for backward-compatible `__setstate__` in general. 
1. First of all, to achieve that, you'll need to execute some part of `__init__` inside, which will lead to performance issues (as `__init__` may contain heavy converters/validators). 
2. Some attributes may have non-idempotent converter (i.e. `converter(converter(value)) != converter(value)`), which will lead to unpicklable classes
3. Some attributes may be actually some kind of cached properties, having `init=False` and default factory depending on some other attributes, which may lead to.

**Comment by hynek** [comment]
I'm sympathetic to the problem, but as far as I can see (as as far as @Drino has elaborated), attrs's pickling works fine and we have AFAICT exhaustive tests to prove that:

https://github.com/python-attrs/attrs/blob/6b84ad2db26fc710577499c6c75ccaac08a263e1/tests/test_functional.py#L272-L322

ISTM that ruamel.yaml is making some invalid assumptions about how exactly pickling works and what guarantees exist? Or can you provide a failing test case that uses `pickle`?

Since ruamel.yaml seems to be maintained, maybe this is fixable in other ways? Adding unnecessary methods has downsides for _everyone_ and this is the first time, this problem has popped up.

**Comment by wsanchez** [comment]
Summarizing:
 * `ruamel.yaml` has some issues with `attrs`.
 * The assertion was that `attrs` is doing something incorrect WRT interfaces intended for pickling and are confident that this is not the case.
 * We think `ruamel.yaml` is making some incorrect assumptions about the pickling interfaces.

So I'm going to tag this as a bug, since that's the assertion made, but then I'm tempted to close as invalid, absent some additional information to the contrary.

Alternatively, there's a feature request in here, but it sounds like making `__getstate__` and `__setstate__` comply with `ruamel.yaml`'s expectations is a non-starter here.  Yeah?

**Comment by TMiguelT** [comment]
I think that's accurate. I think it's increasingly clear that it's `ruamel.yaml` that's misusing its interfaces and not `attrs`. 

To summarise, `ruamel.yaml` assumes this is sufficient for deserialising a dict class:
```python
data = cls.__new__(cls)
data.__dict__.update(state)
```
This is not the case for an `attrs` class, because `attrs` requires that `__init__()` is also called. Alternatively, if you have a `__setstate__()` defined (ie because you have used `@attr.s(getstate_setstate=True)`, `ruamel.yaml` instead tries:
```python
data = cls.__new__(cls)
data.__setstate__(some_dict)
```
I believe this works fine, as long as you also _dumped_ the class instance when it had `getstate_setstate=True`. I did not, which caused a failure, but that's on me and not a fault of either library.

This issue could be fairly trivially solved in `attrs` by implementing a `to_yaml`/`from_yaml` method, but this is non-standard so I understand why you probably don't want this. The best path forward is probably adding a feature to `ruamel.yaml` that will allow the `__init__` to be called.

**Comment by wsanchez** [comment]
OK, going to close.

**Comment by AvdN** [comment]
I have tried to get this to worked based on [this bug report](https://sourceforge.net/p/ruamel-yaml/tickets/394/).

I cannot do `__init__()` *instead* of `__new__()` for several reasons, but most  of all because at the time of doing the `__new__` call I cannot have all the parameters necessary as recursion is allowed in YAML. 

What I do now (ruamel.yaml 0.17.13) is test for the `__attrs_attrs__` attribute and if there call `data.__init__(state)` instead of `.update()`. This seems to work:

    import sys
    import attr
    import ruamel.yaml

    @attr.s()
    class A:
        yaml_tag = '!A'
        a = attr.ib(init=True, kw_only=True)
        b = attr.ib(init=True, kw_only=True, default=2)

    s = """
    - !A
        a: 1
    """

    yaml = ruamel.yaml.YAML()
    yaml.register_class(A)

    y = yaml.load(s)
    print(y)

    print(A(a=1))
    yaml.dump(y, sys.stdout)

which gives:

    [A(a=1, b=2)]
    A(a=1, b=2)
    - !A
      a: 1
      b: 2



But has not been tested beyond this example (I haven't used the `attrs` package) So please file bug reports (with complete code that recreates the problem, like that which ssph provided) if you find anything that doesn't work and I'll have a look.

**Comment by hynek** [comment]
That’s great news – thank you Anthon!

**Comment by multimeric** [comment]
Hi @AvdN, are you saying that you have updated `ruamel.yaml` itself to do this behaviour automatically? Because that's great if so. Although I wonder if there's a more extensible way you can tell `ruamel` to call `__init()__` rather than just fixing this for `attrs` specifically.

**Comment by AvdN** [comment]
@multimeric Yes that is what I did. I could add a parameter to registering classes, or have an alternative for classes that have defaults for all their parameters, but there are all kind of restrictions (like each parameter name in `__init__` must exactly correspond to the attribute name, as ruamel.yaml, if no explicit representation method is provide, dumps attribute names). And then there are various ways to register classes, so they might all need alternatives.

So initially I want to leave it like is, and not change the way things are done too much, that also gives a higher chance of compatibility with what PyYAML is doing (i.e. AFAICT the whole `__new__()` / `.update()` stuff is pre 2006 code that I forked by necessity)

So just file bugs (or post on StackOverflow) if there are problems/questions and I'll try to adapt as necessary.

And of course someone can just do `someinstance.__attrs_attrs__` to have ruamel.yaml call `__init__()`, if they really don't want to implement the represent method on their class as they should.

**Comment by multimeric** [comment]
> I could add a parameter to registering classes

I like this idea? How about you accept an "initializer" function e.g. so you can do it trivially for `attrs`:

```python
yaml.register_class(A, init=lambda attrs, obj: obj.__init__(**kwargs))
```

Then more complex classes can make a custom init function:
```python
def init_prefix(attrs: dict, obj: A):
    for key, value in attrs.items():
        setattr(obj, f"prefix_{key}", value)

yaml.register_class(A, init=init_prefix)
```

Then you can keep the default behaviour the same (using `__new__` + `update()`)  and only use this when users opt in.
